notes/overriding.md

```diff
@@ -0,0 +1,675 @@
+reference to the previous layer and the `self` reference to the final value.
+
+Take two consecutive transformations:
+
+1. Set the `a` field to `1`
+2. Set the `b` field to `1`, add `1` to the `a` field
+
+```nix
+let baseRepr = self: {c = self.a + self.b;}; in
+let overlay1 = self: super: {a = 1;}; in
+let overlay2 = self: super: {b = 1; a = super.a + 1;}; in
+let applyOverlays = self:
+  let base = baseRepr self; in
+  let first = base // overlay1 self base; in
+  let second = first // overlay2 self first; in
+  second; in
+let fixpoint = applyOverlays fixpoint; in fixpoint
+```
+
+In practice, the `super // ..` and fixpoints parts can be factorised in
+dedicated helper functions.
+
+**limits**
+ - Modifications must be applied in the right order.
+ - Still need to manipulate representations.
+ - Overriding nested fields is still clumsy:
+   ```nix
+   self: super:
+   {
+     lib = (super.lib or {}) // {
+       firefoxVersion = ...;
+     };
+   }
+   ```
+
+### NixOs module system
+
+The NixOS module system takes a different approach. There, all the basic blocks
+-- the modules -- are merged together in an unspecified order. What's deciding
+the priority of one option over the other are attributes that are explicitly
+stated in the modules themselves.
+
+Nonetheless, modules need to combine different pieces of data together rather
+than just overriding one value with another. This is made difficult by the
+disappearance of the explicit `super`, compared to overlays.
+
+The issue is solved by custom merge functions, that can define how to combine
+different values of the field of a configuration. By default, when merging two
+list values, the module system only knows how to replace one value with the
+other, because there's no canonical and commutative way of merging two lists.
+However, the user can specify that these lists should in fact be concatenated,
+resulting in the possibility of defining a list of paths by pieces:
+
+```nix
+# Some module
+{
+  paths = mkOption {
+    type = types.listOf types.path;
+  };
+}
+
+# Config
+[..]
+someModule.paths = [foo/bar];
+
+# Other config
+[..]
+someModule.paths = [/bar/baz];
+
+# Resulting config: [foo/bar /bar/baz]
+```
+
+## Overriding elsewhere
+
+To my knowledge, among the other configuration languages Nickel is close to,
+Jsonnet is the only one to feature a proper overriding mechanism.
+
+### CUE
+CUE allows a form of late-binding for recursive attributes:
+
+```
+$cat test.cue
+fields: {
+ a: int
+ b: a + 1
+} & {
+ a: 2
+}
+$cue eval test.cue
+fields: {
+    a: 2
+    b: 3
+}
+```
+
+However, by the very nature of CUE's merge operation, the value of `a` can't be
+changed by merging, it can only be made more precise:
+
+```
+$cat test2.cue
+fields: {
+ a: 1
+ b: a + 1
+} & {
+ a: 2
+}
+$cue eval test2.cue
+fields.a: conflicting values 1 and 2:
+    ./test2.cue:1:9
+    ./test2.cue:2:5
+    ./test2.cue:5:5
+```
+
+It is thus not really an overriding mechanism.
```

Comment:
Well, it kind of is. With the default value thing. But it is not as complete as what we are trying to achieve, probably. There lacks in the initial paragraph an explicit goal that would rule Cue's mechanism out.

Reply:
My bad, I forgot that CUE also had default values. Then the limitation is the one mentioned in the section on priorities, that you probably don't want to have the config writers to annotate everything with `default` to make it overridable. But indeed, it is an overriding mechanism. I'll rewrite this bit